interpreter.go

```diff
@@ -454,65 +492,76 @@ func (i *interpreter) evaluate(a ast.Node, tc tailCallStatus) (value, error) {
 		return sb.self, nil
 
 	case *ast.Var:
-		return e.evaluateTailCall(e.lookUpVar(node.Id), tc)
+		foo := i.stack.lookUpVarOrPanic(node.Id)
+		return foo.getValue(i, trace)
 
 	case *ast.SuperIndex:
-		index, err := e.evalInCurrentContext(node.Index, nonTailCall)
+		index, err := i.evaluate(node.Index, nonTailCall)
 		if err != nil {
 			return nil, err
 		}
-		indexStr, err := e.getString(index)
+		indexStr, err := i.getString(index, trace)
 		if err != nil {
 			return nil, err
 		}
-		return objectIndex(e, i.stack.getSelfBinding().super(), indexStr.getString())
+		return objectIndex(i, trace, i.stack.getSelfBinding().super(), indexStr.getString())
 
 	case *ast.InSuper:
-		index, err := e.evalInCurrentContext(node.Index, nonTailCall)
+		index, err := i.evaluate(node.Index, nonTailCall)
 		if err != nil {
 			return nil, err
 		}
-		indexStr, err := e.getString(index)
+		indexStr, err := i.getString(index, trace)
 		if err != nil {
 			return nil, err
 		}
-		field := tryObjectIndex(i.stack.getSelfBinding().super(), indexStr.getString(), withHidden)
-		return makeValueBoolean(field != nil), nil
+		hasField := objectHasField(i.stack.getSelfBinding().super(), indexStr.getString(), withHidden)
+		return makeValueBoolean(hasField), nil
 
 	case *ast.Function:
 		return &valueFunction{
-			ec: makeClosure(i.getCurrentEnv(a), node),
+			ec: makeClosure(i.stack.getCurrentEnv(a), node),
 		}, nil
 
 	case *ast.Apply:
 		// Eval target
-		target, err := e.evalInCurrentContext(node.Target, nonTailCall)
+		target, err := i.evaluate(node.Target, nonTailCall)
 		if err != nil {
 			return nil, err
 		}
-		function, err := e.getFunction(target)
+		function, err := i.getFunction(target, trace)
 		if err != nil {
 			return nil, err
 		}
 
 		// environment in which we can evaluate arguments
-		argEnv := i.getCurrentEnv(a)
+		argEnv := i.stack.getCurrentEnv(a)
 		arguments := callArguments{
-			positional: make([]potentialValue, len(node.Arguments.Positional)),
+			positional: make([]*cachedThunk, len(node.Arguments.Positional)),
 			named:      make([]namedCallArgument, len(node.Arguments.Named)),
 			tailstrict: node.TailStrict,
```

Comment:
More generally, evaluating needs to return some continuation instead of actually recursively invoking itself.  That allows you to come out before you go back in again, but at the expense of heap allocation.  I don't think you have to build long chains though.  It should be possible to keep going in and out without building up a long chain, right?  The caller just has to force the continuation as soon as it returns.

Reply:
Ok, I think I know exactly how I would implement it, and the diff should be pretty small. So I can just do it, once this change is merged.